### ISSUE TITLE
Regression in unpacking array of scalars

### ISSUE THREAD

**Issue by LukasBanana** [KhronosGroup/SPIRV-Cross#1172]
We stumbled across an issue with arrays of scalars in SPIRV-Cross that looks like a regression in the unpacking mechanism (see example on [shader-playground.timjones.io](http://shader-playground.timjones.io/48c302dc5be75a319c22de743745fd80)). This also reproduces with the latest revision of the master branch.

The following HLSL example casts the element of an array of scalars to a `float4`.
```hlsl
float ScalarValues[1];
float4 Main() : SV_Position
{
    return (float4)ScalarValues[0];
}
```
The output is valid Metal code but produces a wrong result:
```metal
#include <metal_stdlib>
#include <simd/simd.h>
using namespace metal;
struct type_Globals
{
    float4 ScalarValues[1];
};
struct Main_out
{
    float4 gl_Position [[position]];
};
vertex Main_out Main(constant type_Globals& _Globals [[buffer(0)]])
{
    Main_out out = {};
    out.gl_Position = float4(_Globals.ScalarValues[0]); // <--- .x swizzle missing
    return out;
}
```
I'm not 100% sure if the actual issue is the translation from an array of scalars to an array of vectors, but it might be okay due to implicit padding in HLSL.
That said I think arrays of scalars should not have padding in HLSL.
Moreover, translating to GLSL does not introduce an array of vectors.

Alternatively, `_Globals.ScalarValues[0]` should be extended by a `.x` swizzle operator or otherwise potentially zero initialized values are read.
In other words, either `float4(_Globals.ScalarValues[0].x)` or `_Globals.ScalarValues[0].xxxx` must be emitted.

Kind regards
Lukas

**Comment by HansKristian-Work** [comment]
Looks like a regression indeed, sorry about that. MSL needs a float4 array workaround to deal with std140 UBO/cbuffer layout.
